n3fit/src/evolven3fit_new/utils.py

```diff
@@ -0,0 +1,179 @@
+    not provided.
+    """
+    if Qfin is None:
+        if Q_points is None:
+            return (
+                np.array(
+                    [
+                        1.6500000e00,
+                        1.7874388e00,
+                        1.9429053e00,
+                        2.1193749e00,
+                        2.3204100e00,
+                        2.5502944e00,
+                        2.8142025e00,
+                        3.1184122e00,
+                        3.4705775e00,
+                        3.8800751e00,
+                        4.3584516e00,
+                        4.9200000e00,
+                        4.9200000e00,
+                        5.5493622e00,
+                        6.2897452e00,
+                        7.1650687e00,
+                        8.2052867e00,
+                        9.4481248e00,
+                        1.0941378e01,
+                        1.2745972e01,
+                        1.4940062e01,
+                        1.7624572e01,
+                        2.0930715e01,
+                        2.5030298e01,
+                        3.0149928e01,
+                        3.6590777e01,
+                        4.4756282e01,
+                        5.5191298e01,
+                        6.8637940e01,
+                        8.6115921e01,
+                        1.0903923e02,
+                        1.3938725e02,
+                        1.7995815e02,
+                        2.3474820e02,
+                        3.0952544e02,
+                        4.1270732e02,
+                        5.5671861e02,
+                        7.6011795e02,
+                        1.0509694e03,
+                        1.4722574e03,
+                        2.0906996e03,
+                        3.0112909e03,
+                        4.4016501e03,
+                        6.5333918e03,
+                        9.8535186e03,
+                        1.5109614e04,
+                        2.3573066e04,
+                        3.7444017e04,
+                        6.0599320e04,
+                        1.0000000e05,
+                    ]
+                )
+                ** 2
+            )
+        else:
+            raise ValueError(
+                "q_fin and q_points must be specified either both or none of them"
+            )
+    elif Q_points is None:
+        raise ValueError(
+            "q_fin and q_points must be specified either both or none of them"
+        )
+    else:
+        grids = []
+        Q_ini = Q0
+        num_points_list = []
+        for masses in match_dict.keys():
+            match_scale = masses * match_dict[masses]
+            num_points = int(
+                Q_points * (np.log(match_scale / Q0) / np.log(Qfin / Q_ini))
+            )
+            num_points_list.append(num_points)
+            grids.append(np.geomspace(Q0**2, match_scale**2, num=num_points))
+            Q0 = match_scale
```

Comment:
I'm not sure I would use `Q0` as name for the starting scale of each patch, since we usual name as `Q0` a very specific thing (the fitting scale).
Consider using another variable with a different name, and initialize it with `Q0`.